File: 496.py
class Solution:
    def nextGreaterElement(self, nums1: List[int], nums2: List[int]) -> List[int]:
        # time: O(m + n), space: O(m + n)
        res = []
        for n in nums1:
            idx = nums2.index(n)
            greaterElem = 0
            print(idx)
            for i in range(idx + 1, len(nums2)):
                print(i)
                if nums2[i] > n:
                    greaterElem = nums2[i]
                    break
            res.append(greaterElem if greaterElem else -1)
        return res

File: 496.java
class Solution {
    public int[] nextGreaterElement(int[] nums1, int[] nums2) {
        // time: O(n), space: O(n)
        int[] res = new int[nums1.length];
        Stack<Integer> st = new Stack<>();
        HashMap<Integer, Integer> map = new HashMap<>();
        for (int n : nums2) {
            while (!st.isEmpty() && n > st.peek()) {
                map.put(st.pop(), n);
            }
            st.add(n);
        }
        int i = 0;
        for (int n : nums1) {
            res[i++] = map.getOrDefault(n, -1);
        }
        return res;
    }
}